CONTRIBUTING.md

```diff
@@ -0,0 +1,278 @@
+   only of collection types which may be empty it should derive `Default` trait.
+
+3. **Error types** (both structs and enums) must implement `Display` and `Error`
+   traits manually, and should provide `Error::source` function if some of the
+   error cases contain other error type.
+
+4. `Display` should be implemented for all data types which may be presented to
+   the end user (not developers!), for instance in command line or as a part of
+   GUI. Here are some guidelines: 
+   - Normally, `Display` implementation should not just repeat `Debug` and
+     structure the data in some visually-acceptable way. 
+   - One should pay attention to the ability of providing alternative ways of 
+     data formatting with `{:#}` formatting string option, detectable by 
+     `std::fmt::Formatter::alternate()` function. Other important options to
+     look at are `align`, `fill`, `pad`, `precision` and `width`. 
+   - When displaying the member fields it is important to consider the ability
+     to pass them display formatting options; thus, 
+     `Display::fmt(&self.field, f)?;` is preferable over 
+     `write!(f, "{}", self.field)?;`
+
+5. Serde serializers should be implemented for all data types which may persist
+   or may be presented in the UI or API as JSON/YAML and other kinds of data 
+   representations (in fact, these are all data types).
+
+The discussion about trait derivation can be read at 
+[the tracking issue](https://github.com/rust-bitcoin/rust-bitcoin/issues/555).
+
+### MSRV
+
+The Minimal Supported Rust Version (MSRV) is 0.29; it is enforced by our CI. 
+Later we plan to increase MSRV to support Rust 2018 and you are welcome to check
+the [tracking issue](https://github.com/rust-bitcoin/rust-bitcoin/issues/510).
+
+### Naming conventions
+
+Naming of data structures/enums and their fields/variants must follow names used
+in Bitcoin Core, with except to:
+- case, which should follow Rust standards (i.e. PascalCase for types and 
+  snake_case for fields and variants)
+- `C`-prefix, which should be omitted
+
+
+## Security
+
+Security is the primary focus of Rust-LNPBP; disclosure of security
+vulnerabilities helps prevent user loss of funds. If you believe a vulnerability
+may affect other  implementations, please inform them.
+
+Note that Rust-LNPBP is currently considered "pre-production" during this time,
+there is no special handling of security issues. Please simply open an issue on
+Github.
+
+
+## Testing
+
+Related to the security aspect, rust bitcoin developers must be taking testing
+very seriously. Due to the modular nature of the project, writing new functional
+tests is easy and good test coverage of the codebase is an important goal.
+Refactoring the project to enable fine-grained unit testing is also an ongoing
+effort.
+
+Fuzzing is heavily encouraged: feel free to add related material under `fuzz/`
+
+Mutation testing is planned; any contribution there would be warmly welcomed.
+
```

Comment:
Maybe add a requirement to test `unsafe` code with sanitizers including Miri?

Reply:
Do we allow `unsafe` at all here?

Reply:
Added a dedicated section addressing `unsafe`: https://github.com/rust-bitcoin/rust-bitcoin/pull/587/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055R262-R266